### ISSUE TITLE
Porting PostProcessingProfiles from V1 to V2

### ISSUE THREAD

**Issue by AdamFrisby** [Unity-Technologies/PostProcessing#684]
Hi there - 

I'm currently writing some scripts to automate porting between V1 and V2 - while most values have remapped fairly easily; I have two areas (so far) which do not remap cleanly.

The first is the Neutral Tonemapper - in V1, you have control over the White Balance directly - this is not present in V2. I was thinking of using the 'Custom' curve option here to reimplement the older curve. It looks like the curve simulator in the old ColorGradingModelEditor.cs could be the basis of this, but some pointers would be appreciated.

The second area I've run into is the Trackwheels - the older post stack used the Log wheels by default, unfortunately it looks like only the Linear wheels are supported in V2 - is there a simple way to approximate these? (_Super dumb idea_; can I just Log(X) to convert to Linear?)

I'll post the code for these porting utilities when I'm done with them - thankfully it looks like outside of the Tonemapper, all other values can be remapped pretty easily.

**Comment by AdamFrisby** [comment]
Here's the work so far (yes, it doesn't fill in override params yet - need to add that in): https://github.com/AdamFrisby/Post-Stack-Converter/blob/master/PostProcessingV1toV2.cs

**Comment by Chman** [comment]
Hi,

Color Grading is one of the effects that will be hard to port automatically between v1 and v2 (if not impossible). The maths have been rewritten in v2 and thus behave differently and cannot be remapped accurately.

For the tonemapper, we removed the ability to customize the neutral curve (it wouldn't be neutral if you could change it) mostly because the way the settings mapped to curve made it impossible to tweak in a logical way (artists & users feedback). This is why there's a "Custom" tonemapper now with easier and more direct controls of the toe, mid and shoulder parts of the curve. With that said, mapping the neutral curve to the custom one mathematically is going to be challenging and my recommendation would be to do it manually in the inspector, it's fairly easy to eyeball. Quick note: there's no control of the white balance on the neutral tonemapper, only white in/out/clip which are in now way related to an actual white balancing process.

As for the trackwheels, in v1 there were Log and Linear options like you said, and the documentation recommended to use Log when working in HDR and Linear in LDR due to the way the wheels were implemented. This was quite error-prone and we found out most users were confused about the choice of mode. So in v2, you either work in a LDR or an HDR context (first drop down in Color Grading v2). This will define how the input is interpreted and the shader code is different for both. You won't be able to reliably map v1's log wheels to v2's LDR or HDR wheels unfortunately as the maths and the way they're applied are quite different. The linear wheels could be approximated fairly well, but the log ones can't. I unfortunately don't have a good solution to offer you for this specific case.

Most of the other values should map fairly well but be aware that in v1 some happened in log space whereas they are in linear space in v2. A quick glance at the code shows that in v1, `Hue VS Sat`, `Sat VS Sat`, `Lum VS Sat` and `Saturation` are applied on the log input. They're all linear in v2. This will affect your converter and result in a different output. Also, `Channel Mixer` is done earlier in the grading chain now, it used to be the last step before tonemapping.

**Comment by AdamFrisby** [comment]
Thanks @Chman - this is very useful information.

I'm not specifically aiming for a perfect reproduction - 'close enough' is good enough, although the solution I'm implementing does need to be fully automated. I'll see if I can reproduce the old algorithm in a custom curve for the tonemapper (I think this is a viable approach, if dodgy) for compatibility purposes.

It looks like the log trackwheels will be the biggest obstacle - if there's no good solutions, do you have a bad one?

**Comment by AdamFrisby** [comment]
Yeah, I understand dropping compat (although I am hoping this might be the last time for a while...)

Our use case is 'interesting' (according to Enterprise Support, we're "An Edge Case") - we run a multi-user collaborative 3D virtual world, where developers can upload scenes & content built in Unity into the world itself. Our goal is to make sure content artists have made in the past continues to work in the future, without their help. 

They can of course, go back and refine things - but we have to assume they wont. We've got some room for changes which change visuals - although we try minimise that wherever possible; but the key is any upgraders do need to be fully automated as we've got literally terabytes of content and us doing any hand tweaking is out of the question.

If you're curious - take a peek at https://developer.space (little out of date now) and https://sine.space - it's quite cool being able to upload stuff into the world from Unity. :)

**Comment by AdamFrisby** [comment]
I will caveat my responses above that I know it is possible to write a custom effect in the PPV2 and we could port the tone-mapper wholesale (I am not ruling this possibility out) - but I would prefer not to go down that path as it forces us to support it.

**Comment by Chman** [comment]
Edge case indeed, but an interesting one nonetheless.

Have you thought about not porting v1 to v2 and rather support both? I haven't checked in some time but v1 and v2 should be able to live in the same project, meaning you could potentially use v2 for new content and still use v1 for "legacy" content. There's probably some side effects I haven't thought about because I'm not too familiar with your product or it may also be a very stupid idea, but... Just throwing the idea around.

**Comment by AdamFrisby** [comment]
Yeah; I have noticed they both sit within the project without breaking (appreciated - thankyou!) and I have debated preserving both as well, but I think that might be asking for trouble if people start making content that uses the newer volumes, and adds it to a scene with the older profiles.

I think my current strategy is going to be play with the colour grading module a bit - see if we can get something that's "good enough" automatically, if that fails, then we'll copy the colour grading module. I am setting myself up for a fall ... but I'm betting across the common usage ranges we can find a point where the linear/log conversion mostly works.

**Comment by Chman** [comment]
Just thought about something. Assuming your converter runs in Unity there might be a simpler way to convert color grading from v1 to v2. Use the v1 profile to generate a LUT to a texture, and use this LUT into v2. That way you get a perfect match.

**Comment by AdamFrisby** [comment]
Yeah that is a possibility, it's not -quite- perfect since we lose our blending if someone starts adding volumes, but it may be _a_ solution.

I hope you don't mind me asking all these questions - I've run into another issue, which could be tonemapper or bloom related.

Bloom:
Diffusion - this appears to be the old 'radius' parameter, but I'm not sure if it's switched to radius from diameter - the values feel 'off' by a clean 2x. (I'm also noticing it's a bit more spherical now than exponential in the blurring?)
Intensity - this seems off by a factor of 10x from before 

This _could_ be the tonemapper again though.

**Comment by AdamFrisby** [comment]
Well, with some slight tweaks; it's -mostly- working - the big change I can still see is it seems like bloom is less saturated than before (actually saturation seems lower across the board), I think this might be still some errors in the conversion of the trackball data (I think there's some blood left in that stone...)

Before:
![image](https://user-images.githubusercontent.com/114041/47409518-b671e900-d7ae-11e8-95d3-5241a07cd53e.png)

After:
![image](https://user-images.githubusercontent.com/114041/47409503-abb75400-d7ae-11e8-8264-326892889a21.png)

**Comment by Chman** [comment]
Hey,

Regarding bloom, radius is indeed diffusion. There shouldn't much difference between the two (except in v2 you can go higher than in v1), the problem you're seeing is most likely due to how you deal with intensity. In v1, intensity was a simple multiplier. In v2 it's exponential: `2^(intensity / 10) - 1` so you'll have to remap it.

Tonemapping differences will have some effect on the final comparison as well but it should be manageable as you found out.

**Comment by AdamFrisby** [comment]
Thanks for this - very useful!

One last question (for the moment) am I correct in that enabling tonemapping in the PostV1 stack forced LDR output? (I can't remember - there's been a mix of tonemapping components over the years)

**Comment by AdamFrisby** [comment]
Great, thanks - some of the older effects I've encountered were using the YRGB curves, which only work in V2 in the LDR mode.

**Comment by rhys-vdw** [comment]
Thanks @AdamFrisby, this is great. I would suggest that the maintainers add this script to documentation. Saved me a lot of time.

**Comment by AdamFrisby** [comment]
@Chman Hey - thanks for your help earlier, I have in the end gone down the road of using the existing code to generate a LUT - I'm using the older bakedLut property stuffed into the .model inside ColorGradingComponent; however I am finding it is way too dark.

I'm suspecting there is some linear/gamma conversion going on here - but I've been trying to manually correct that and it's still not right. Any clues/ideas?

**Comment by JashanChittesh** [comment]
@AdamFrisby Have you tried your conversion script in Unity 2018.3(.5)? I noticed that when closing and re-opening the editor, the converted profiles are all empty, which is due to Unity not writing the overrides down to the files. So they show up in the editor - as long as they are still in memory. See also: [Generated assets are all empty #1](https://github.com/AdamFrisby/Post-Stack-Converter/issues/1).

**Comment by AdamFrisby** [comment]
Hey @JashanChittesh - I have, I use 2018.3 - we're doing the conversion at runtime, so the writing to disk isn't strictly required for our use case.

That said, try calling EditorUtility.SetDirty on it - might force it to flush next time you call AssetDatabase.SaveAssets()

**Comment by JashanChittesh** [comment]
Ah, ok, runtime conversion does indeed change the story. The odd thing is that neither SetDirty, nor Undo.RecordObject(...) fixed this. I could even first create the new asset, then load it, then make the changes, and it still didn't save.

Looking at the custom inspector solved the issue: [EffectListEditor.cs](https://github.com/Unity-Technologies/PostProcessing/blob/v2/PostProcessing/Editor/EffectListEditor.cs); in AddEffectOverride(...), it says: 

`AssetDatabase.AddObjectToAsset(effect, asset);`

... and that does the trick. We only have to do this for Bloom and ColorGrading, the full solution would be adding it everywhere (we do first save the asset, and that might be required for that approach, but that's easy enough to do).

**Comment by AdamFrisby** [comment]
Feel free to open up a pull request with that change. I'll merge it back in - only one request, please #ifdef UNITY_EDITOR that section, as AssetDatabase.* won't exist at runtime.
